cpp/ql/test/query-tests/Security/CWE/CWE-242/semmle/tests/tests.cpp

```diff
@@ -362,4 +362,17 @@ void test6(unsigned unsigned_value, int value) {
 
 	sprintf(buffer4, "%d", (int)c); // BAD: e.g. -127 does not fit
 	sprintf(buffer5, "%d", (int)c); // GOOD: -127..128 fits
-}
\ No newline at end of file
+}
+
+void test7() {
+    char buffer[2];
+
+	for (int i = 0; i <= 11; i++) {
+		if (0 <= i && i <= 9) {
+			sprintf(buffer, "%d", i - 0);  // GOOD, 0 <= i <= 9
+		}
+		if (1 <= i && i <= 10) {
+			sprintf(buffer, "%d", i - 1);  // GOOD, 0 <= i - 1 <= 9
+		}
```

Comment:
I'm not quite sure what the problem is.  Did you mean to show a false positive case?

Reply:
yes this is a weird FP. What seems weird to me is that `i - 0` has a different range being found out than `i` alone (removing `- 0` makes range analysis find the correct range here), or that the analysis works correctly also if we do
```cpp
void test7(int i) {
  char buffer[2];
  if (0 <= i && i <= 11) {
    if (i <= 9) {
      sprintf(...);
```